apollo-router/src/json_ext.rs

```diff
@@ -192,6 +198,53 @@ impl ValueExt for Value {
         }
     }
 
+    fn type_aware_deep_merge(&mut self, other: Self, schema: &Schema) {
+        match (self, other) {
+            (Value::Object(a), Value::Object(b)) => {
+                for (key, value) in b.into_iter() {
+                    let k = key.clone();
+                    match a.entry(key) {
+                        Entry::Vacant(e) => {
+                            e.insert(value);
+                        }
+                        Entry::Occupied(e) => match (e.into_mut(), value) {
+                            (Value::String(type1), Value::String(type2))
+                                if k.as_str() == TYPENAME =>
```

Comment:
If you change this line to
```suggestion
                                if e.key().as_str() == TYPENAME =>
```
you can avoid the `key.clone()` above, which avoids some unnecessary metadata cloning in the common case.